File: dotnet/ArchiveDir.cs
using System;
using System.IO;
using System.Data.SQLite;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using System.Drawing;
using System.Drawing.Imaging;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using SkiaSharp;
using MediaToolkit;
using MediaToolkit.Model;
using MediaToolkit.Options;

namespace ArchiveDir
{
    class Program
    {
        static void Main(string[] args)
        {
            // Load environment variables from .env file if needed
            // using DotNetEnv;

            // DotNetEnv.Env.Load();

            // Parse command-line arguments
            var parser = new ArgumentParser(args);
            string srcDirectory = parser.GetValue("--src_directory") ?? Environment.GetEnvironmentVariable("SOURCE_DIRECTORY");
            string destDirectory = parser.GetValue("--dest_directory") ?? Environment.GetEnvironmentVariable("DESTINATION_DIRECTORY");
            string dbPath = parser.GetValue("--db_path") ?? Environment.GetEnvironmentVariable("DB_PATH");
            string azureContainer = parser.GetValue("--azure_container") ?? Environment.GetEnvironmentVariable("AZURE_CONTAINER");
            string azureConnectionString = parser.GetValue("--azure_connection_string") ?? Environment.GetEnvironmentVariable("AZURE_CONNECTION_STRING");

            DateTime startTime = DateTime.Now;
            Console.WriteLine($"Process started at: {startTime}");

            long originalSize = GetDirectorySize(srcDirectory);

            // Generate a timestamp
            string timestamp = DateTime.Now.ToString("yyyyMMddHHmmss");

            // Open a single database connection
            string connectionString = $"Data Source={dbPath};Version=3;";
            using (var conn = new SQLiteConnection(connectionString))
            {
                conn.Open();

                // Redirect warnings or logs if necessary
                // Implement custom warning handling if required

                try
                {
                    // Compress files and save metadata to the database
                    CompressFilesAndSaveToDb(srcDirectory, destDirectory, conn, timestamp);

                    long compressedSize = GetDirectorySize(destDirectory);

                    DateTime endTime = DateTime.Now;
                    Console.WriteLine($"Process ended at: {endTime}");
                    Console.WriteLine($"Total duration: {endTime - startTime}");
                    Console.WriteLine($"Total original size: {FormatSize(originalSize)}");
                    Console.WriteLine($"Total compressed size: {FormatSize(compressedSize)}");

                    Console.WriteLine($"Directory tree saved to database: {dbPath}");

                    // Upload to Azure Blob Storage if specified
                    if (!string.IsNullOrEmpty(azureContainer) && !string.IsNullOrEmpty(azureConnectionString))
                    {
                        UploadToAzure(azureContainer, azureConnectionString, destDirectory, timestamp, conn);
                        Console.WriteLine($"Files uploaded to Azure Blob Storage container: {azureContainer}");

                        // Remove the created destination files and directory
                        RemoveDirectory(destDirectory);
                        Console.WriteLine($"Removed destination directory: {destDirectory}");
                    }
                }
                finally
                {
                    conn.Close();
                }
            }
        }

        static long GetDirectorySize(string directory)
        {
            long totalSize = 0;
            foreach (string file in Directory.EnumerateFiles(directory, "*", SearchOption.AllDirectories))
            {
                FileInfo fi = new FileInfo(file);
                totalSize += fi.Length;
            }
            return totalSize;
        }

        static string FormatSize(long size)
        {
            string[] units = { "B", "KB", "MB", "GB", "TB" };
            double sizeD = size;
            int unit = 0;
            while (sizeD >= 1024 && unit < units.Length - 1)
            {
                sizeD /= 1024;
                unit++;
            }
            return $"{sizeD:F2} {units[unit]}";
        }

        static void CompressFilesAndSaveToDb(string srcDir, string destDir, SQLiteConnection conn, string timestamp)
        {
            var cursor = conn.CreateCommand();

            // Create tables
            cursor.CommandText = @"
                CREATE TABLE IF NOT EXISTS files (
                    id TEXT PRIMARY KEY,
                    filename TEXT,
                    filepath TEXT,
                    content_type TEXT,
                    size INTEGER,
                    creation_time TEXT,
                    modification_time TEXT,
                    thumbnail BLOB,
                    is_duplicate INTEGER,
                    original_path TEXT,
                    batch TEXT
                );
                CREATE TABLE IF NOT EXISTS events (
                    id TEXT PRIMARY KEY,
                    event_type TEXT,
                    file_path TEXT,
                    message TEXT,
                    timestamp TEXT
                );
            ";
            cursor.ExecuteNonQuery();

            var files = Directory.GetFiles(srcDir, "*", SearchOption.AllDirectories);
            int totalItems = files.Length;
            int processedItems = 0;

            foreach (var filePath in files)
            {
                string relativePath = Path.GetRelativePath(srcDir, filePath);
                string destPath = Path.Combine(destDir, relativePath);
                string destDirPath = Path.GetDirectoryName(destPath);

                if (!Directory.Exists(destDirPath))
                {
                    Directory.CreateDirectory(destDirPath);
                }

                try
                {
                    // Check for duplicates in the database
                    FileInfo fileInfo = new FileInfo(filePath);
                    string originalFilename = fileInfo.Name;
                    string contentType = MimeMapping.GetMimeMapping(filePath);
                    long size = fileInfo.Length;
                    string creationTime = fileInfo.CreationTimeUtc.ToString("o");
                    string modificationTime = fileInfo.LastWriteTimeUtc.ToString("o");

                    var checkDuplicateCmd = conn.CreateCommand();
                    checkDuplicateCmd.CommandText = @"
                        SELECT filepath FROM files WHERE filename = @filename AND size = @size AND creation_time = @creation_time AND modification_time = @modification_time
                    ";
                    checkDuplicateCmd.Parameters.AddWithValue("@filename", originalFilename);
                    checkDuplicateCmd.Parameters.AddWithValue("@size", size);
                    checkDuplicateCmd.Parameters.AddWithValue("@creation_time", creationTime);
                    checkDuplicateCmd.Parameters.AddWithValue("@modification_time", modificationTime);
                    var originalRecord = checkDuplicateCmd.ExecuteScalar();
                    bool isDuplicate = originalRecord != null;

                    if (isDuplicate)
                    {
                        string originalPath = originalRecord.ToString();
                        var insertDuplicateCmd = conn.CreateCommand();
                        insertDuplicateCmd.CommandText = @"
                            INSERT INTO files (id, filename, filepath, content_type, size, creation_time, modification_time, thumbnail, is_duplicate, original_path, batch)
                            VALUES (@id, @filename, @filepath, @content_type, @size, @creation_time, @modification_time, @thumbnail, @is_duplicate, @original_path, @batch)
                        ";
                        insertDuplicateCmd.Parameters.AddWithValue("@id", Guid.NewGuid().ToString());
                        insertDuplicateCmd.Parameters.AddWithValue("@filename", originalFilename);
                        insertDuplicateCmd.Parameters.AddWithValue("@filepath", (relativePath + ".7z").Replace("\\", "/"));
                        insertDuplicateCmd.Parameters.AddWithValue("@content_type", contentType);
                        insertDuplicateCmd.Parameters.AddWithValue("@size", size);
                        insertDuplicateCmd.Parameters.AddWithValue("@creation_time", creationTime);
                        insertDuplicateCmd.Parameters.AddWithValue("@modification_time", modificationTime);
                        insertDuplicateCmd.Parameters.AddWithValue("@thumbnail", DBNull.Value);
                        insertDuplicateCmd.Parameters.AddWithValue("@is_duplicate", 1);
                        insertDuplicateCmd.Parameters.AddWithValue("@original_path", originalPath);
                        insertDuplicateCmd.Parameters.AddWithValue("@batch", timestamp);
                        insertDuplicateCmd.ExecuteNonQuery();

                        processedItems++;
                        continue;
                    }

                    // Compress the file with fastest compression
                    string zipFilePath = destPath + ".7z";
                    CompressFile(filePath, zipFilePath);

                    // Generate thumbnail
                    byte[] thumbnail = null;
                    if (contentType.StartsWith("image/") || contentType.StartsWith("video/"))
                    {
                        thumbnail = GenerateThumbnail(filePath, contentType);
                    }

                    // Save file record to database
                    var insertFileCmd = conn.CreateCommand();
                    insertFileCmd.CommandText = @"
                        INSERT INTO files (id, filename, filepath, content_type, size, creation_time, modification_time, thumbnail, is_duplicate, original_path, batch)
                        VALUES (@id, @filename, @filepath, @content_type, @size, @creation_time, @modification_time, @thumbnail, @is_duplicate, @original_path, @batch)
                    ";
                    insertFileCmd.Parameters.AddWithValue("@id", Guid.NewGuid().ToString());
                    insertFileCmd.Parameters.AddWithValue("@filename", originalFilename);
                    insertFileCmd.Parameters.AddWithValue("@filepath", (relativePath + ".7z").Replace("\\", "/"));
                    insertFileCmd.Parameters.AddWithValue("@content_type", contentType);
                    insertFileCmd.Parameters.AddWithValue("@size", size);
                    insertFileCmd.Parameters.AddWithValue("@creation_time", creationTime);
                    insertFileCmd.Parameters.AddWithValue("@modification_time", modificationTime);
                    insertFileCmd.Parameters.AddWithValue("@thumbnail", thumbnail ?? (object)DBNull.Value);
                    insertFileCmd.Parameters.AddWithValue("@is_duplicate", 0);
                    insertFileCmd.Parameters.AddWithValue("@original_path", filePath);
                    insertFileCmd.Parameters.AddWithValue("@batch", timestamp);
                    insertFileCmd.ExecuteNonQuery();

                    processedItems++;
                    Console.WriteLine($"Processed {processedItems}/{totalItems}: {filePath}");
                }
                catch (Exception e)
                {
                    LogEvent("Error processing file", filePath, e.Message, conn);
                }
            }
        }

        static void CompressFile(string filePath, string zipFilePath)
        {
            string compressionLevel = Environment.GetEnvironmentVariable("COMPRESSION_LEVEL") ?? "1";
            var processInfo = new ProcessStartInfo
            {
                FileName = "7z",
                Arguments = $"a -t7z -mx={compressionLevel} \"{zipFilePath}\" \"{filePath}\"",
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            var process = Process.Start(processInfo);
            process.WaitForExit();
            if (process.ExitCode != 0)
            {
                throw new Exception($"7z compression failed: {process.StandardError.ReadToEnd()}");
            }
        }

        static byte[] GenerateThumbnail(string filePath, string contentType)
        {
            try
            {
                if (contentType.StartsWith("image/"))
                {
                    using (var input = File.OpenRead(filePath))
                    using (var original = SKBitmap.Decode(input))
                    {
                        int width = 128;
                        int height = 128;
                        var resizeInfo = new SKImageInfo(width, height);
                        var resized = original.Resize(resizeInfo, SKFilterQuality.High);
                        using (var image = SKImage.FromBitmap(resized))
                        using (var data = image.Encode(SKEncodedImageFormat.Png, 100))
                        {
                            return data.ToArray();
                        }
                    }
                }
                else if (contentType.StartsWith("video/"))
                {
                    var inputFile = new MediaFile { Filename = filePath };
                    using (var engine = new Engine())
                    {
                        engine.GetMetadata(inputFile);
                        var options = new ConversionOptions { Seek = TimeSpan.FromSeconds(1) };
                        var outputFile = new MediaFile { Filename = Path.GetTempFileName() };
                        engine.GetThumbnail(inputFile, outputFile, options);
                        byte[] thumbnail = File.ReadAllBytes(outputFile.Filename);
                        File.Delete(outputFile.Filename);
                        return thumbnail;
                    }
                }
            }
            catch (Exception)
            {
                // Handle exceptions if needed
            }
            return null;
        }

        static void LogEvent(string eventType, string filePath, string message, SQLiteConnection conn)
        {
            var cmd = conn.CreateCommand();
            cmd.CommandText = @"
                INSERT INTO events (id, event_type, file_path, message, timestamp)
                VALUES (@id, @event_type, @file_path, @message, @timestamp)
            ";
            cmd.Parameters.AddWithValue("@id", Guid.NewGuid().ToString());
            cmd.Parameters.AddWithValue("@event_type", eventType);
            cmd.Parameters.AddWithValue("@file_path", filePath);
            cmd.Parameters.AddWithValue("@message", message);
            cmd.Parameters.AddWithValue("@timestamp", DateTime.Now.ToString("o"));
            cmd.ExecuteNonQuery();
        }

        static void UploadToAzure(string containerName, string connectionString, string destDir, string timestamp, SQLiteConnection conn)
        {
            BlobServiceClient blobServiceClient = new BlobServiceClient(connectionString);
            BlobContainerClient containerClient = blobServiceClient.GetBlobContainerClient(containerName);

            var files = Directory.GetFiles(destDir, "*", SearchOption.AllDirectories);
            int totalFiles = files.Length;
            int uploadedFiles = 0;

            foreach (var file in files)
            {
                string relativePath = Path.GetRelativePath(destDir, file).Replace("\\", "/");
                string blobPath = $"{timestamp}/{relativePath}";
                BlobClient blobClient = containerClient.GetBlobClient(blobPath);

                try
                {
                    Console.WriteLine($"Uploading: {file} -> {blobPath}");
                    using (FileStream fs = File.OpenRead(file))
                    {
                        blobClient.Upload(fs, true);
                    }
                    uploadedFiles++;
                    Console.WriteLine($"Uploaded {uploadedFiles}/{totalFiles}");
                }
                catch (Exception e)
                {
                    LogEvent("Error uploading file to Azure", file, e.Message, conn);
                }
            }
        }

        static void RemoveDirectory(string directory)
        {
            Directory.Delete(directory, true);
        }
    }

    class ArgumentParser
    {
        private readonly Dictionary<string, string> argsDict = new Dictionary<string, string>();

        public ArgumentParser(string[] args)
        {
            string key = null;
            foreach (var arg in args)
            {
                if (arg.StartsWith("--"))
                {
                    key = arg;
                }
                else if (key != null)
                {
                    argsDict[key] = arg;
                    key = null;
                }
            }
        }

        public string GetValue(string key)
        {
            argsDict.TryGetValue(key, out string value);
            return value;
        }
    }
}

File: python/archive_dir.py
import os
import subprocess
import argparse
from datetime import datetime
from tqdm import tqdm
import sqlite3
import mimetypes
import uuid
from azure.storage.blob import BlobServiceClient, BlobClient, ContainerClient
from PIL import Image
from moviepy import VideoFileClip
import cairosvg
import time
import warnings
import sys
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

def get_directory_size(directory):
    total_size = 0
    for dirpath, dirnames, filenames in os.walk(directory):
        for f in filenames:
            fp = os.path.join(dirpath, f)
            total_size += os.path.getsize(fp)
    return total_size

def format_size(size):
    # Convert size to human-readable format
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024:
            return f"{size:.2f} {unit}"
        size /= 1024

def generate_thumbnail(file_path, content_type, conn):
    thumbnail = None
    try:
        if content_type.startswith('image/'):
            if file_path.lower().endswith('.svg'):
                # Convert SVG to PNG
                png_file_path = file_path + '.png'
                cairosvg.svg2png(url=file_path, write_to=png_file_path)
                file_path = png_file_path
            
            with Image.open(file_path) as img:
                img.thumbnail((128, 128))
                thumbnail = img.tobytes()
        elif content_type.startswith('video/'):
            with VideoFileClip(file_path) as video:
                frame = video.get_frame(1)
                img = Image.fromarray(frame)
                img.thumbnail((128, 128))
                thumbnail = img.tobytes()
    except Exception as e:
        log_event("Error generating thumbnail", file_path, str(e), conn)
    return thumbnail

def log_event(event_type, file_path, message, conn):
    retries = 5
    while retries > 0:
        try:
            cursor = conn.cursor()
            cursor.execute('''
                INSERT INTO events (id, event_type, file_path, message, timestamp)
                VALUES (?, ?, ?, ?, ?)
            ''', (str(uuid.uuid4()), event_type, file_path, message, datetime.now().isoformat()))
            conn.commit()
            break
        except sqlite3.OperationalError as e:
            if "database is locked" in str(e):
                retries -= 1
                time.sleep(1)
            else:
                raise

def compress_files_and_save_to_db(src_dir, dest_dir, conn, timestamp):
    cursor = conn.cursor()
    
    # Create tables
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS files (
            id TEXT PRIMARY KEY,
            filename TEXT,
            filepath TEXT,
            content_type TEXT,
            size INTEGER,
            creation_time TEXT,
            modification_time TEXT,
            thumbnail BLOB,
            is_duplicate INTEGER,
            original_path TEXT,
            batch TEXT
        )
    ''')
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS events (
            id TEXT PRIMARY KEY,
            event_type TEXT,
            file_path TEXT,
            message TEXT,
            timestamp TEXT
        )
    ''')
    
    # Count total files and directories for progress bar
    total_items = sum(len(files) + len(dirs) for _, dirs, files in os.walk(src_dir))
    
    with tqdm(total=total_items, desc="Processing items", unit="item") as pbar:
        for root, dirs, files in os.walk(src_dir):
            for dir in dirs:
                dir_path = os.path.join(root, dir)
                relative_path = os.path.relpath(root, src_dir)
                dest_path = os.path.join(dest_dir, relative_path)
                
                if not os.path.exists(dest_path):
                    os.makedirs(dest_path)
                
                # Save directory record to database
                original_filename = dir
                stat = os.stat(dir_path)
                size = stat.st_size
                creation_time = datetime.fromtimestamp(stat.st_ctime).isoformat()
                modification_time = datetime.fromtimestamp(stat.st_mtime).isoformat()
                
                cursor.execute('''
                    INSERT INTO files (id, filename, filepath, content_type, size, creation_time, modification_time, thumbnail, is_duplicate, original_path, batch)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ''', (str(uuid.uuid4()), original_filename, os.path.join(relative_path, dir).replace("\\", "/"), "directory", size, creation_time, modification_time, None, 0, dir_path, timestamp))
                
                pbar.update(1)
            
            for file in files:
                file_path = os.path.join(root, file)
                relative_path = os.path.relpath(root, src_dir)
                dest_path = os.path.join(dest_dir, relative_path)
                
                if not os.path.exists(dest_path):
                    os.makedirs(dest_path)
                
                try:
                    # Check for duplicates in the database
                    original_filename = file
                    content_type, _ = mimetypes.guess_type(file_path)
                    stat = os.stat(file_path)
                    size = stat.st_size
                    creation_time = datetime.fromtimestamp(stat.st_ctime).isoformat()
                    modification_time = datetime.fromtimestamp(stat.st_mtime).isoformat()
                    
                    cursor.execute('''
                        SELECT filepath FROM files WHERE filename = ? AND size = ? AND creation_time = ? AND modification_time = ?
                    ''', (original_filename, size, creation_time, modification_time))
                    original_record = cursor.fetchone()
                    is_duplicate = original_record is not None
                    
                    if is_duplicate:
                        original_path = original_record[0]
                        cursor.execute('''
                            INSERT INTO files (id, filename, filepath, content_type, size, creation_time, modification_time, thumbnail, is_duplicate, original_path, batch)
                            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                        ''', (str(uuid.uuid4()), original_filename, os.path.join(relative_path, file + '.7z').replace("\\", "/"), content_type, size, creation_time, modification_time, None, 1, original_path, timestamp))
                        pbar.update(1)
                        continue
                    
                    # Compress the file with fastest compression
                    zip_file_path = os.path.join(dest_path, file + '.7z')
                    compression_level = os.getenv('COMPRESSION_LEVEL', '5')
                    subprocess.run(['7z', 'a', '-t7z', f'-mx={compression_level}', '-m0=LZMA2', '-md=32m', '-ms=64m', '-mmt=4', '-bd', zip_file_path, file_path], 
                                   stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
                    
                    # Generate thumbnail
                    thumbnail = generate_thumbnail(file_path, content_type, conn) if content_type and (content_type.startswith('image/') or content_type.startswith('video/')) else None
                    
                    # Save file record to database
                    cursor.execute('''
                        INSERT INTO files (id, filename, filepath, content_type, size, creation_time, modification_time, thumbnail, is_duplicate, original_path, batch)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    ''', (str(uuid.uuid4()), original_filename, os.path.join(relative_path, file + '.7z').replace("\\", "/"), content_type, size, creation_time, modification_time, thumbnail, 0, file_path, timestamp))
                    
                    pbar.update(1)
                except Exception as e:
                    log_event("Error processing file", file_path, str(e), conn)
    
    conn.commit()

def upload_to_azure(container_name, connection_string, dest_dir, timestamp, conn):
    blob_service_client = BlobServiceClient.from_connection_string(connection_string)
    container_client = blob_service_client.get_container_client(container_name)
    
    cursor = conn.cursor()
    
    # Count total files for progress bar
    cursor.execute('SELECT COUNT(*) FROM files WHERE batch = ? AND is_duplicate = 0', (timestamp,))
    total_files = cursor.fetchone()[0]
    
    with tqdm(total=total_files, desc="Uploading to Azure", unit="file") as pbar:
        cursor.execute('SELECT filepath FROM files WHERE batch = ? AND is_duplicate = 0 AND content_type != "directory"', (timestamp,))
        for row in cursor.fetchall():
            file_path = row[0]
            # blob_path = os.path.join(timestamp, os.path.relpath(file_path, dest_dir)).replace("\\", "/")

            local_file = (dest_dir + '/' + file_path).replace("\\", "/")
            blob_file = (timestamp + '/' + file_path).replace("\\", "/")
            blob_client = container_client.get_blob_client(blob_file)
            
            try:
                print(f"Uploading: {local_file} -> {blob_file}")
                with open(local_file, "rb") as data:
                    blob_client.upload_blob(data, overwrite=True)
                    pbar.update(1)
            except Exception as e:
                log_event("Error uploading file to Azure", file_path, str(e), conn)

def remove_directory(directory):
    for root, dirs, files in os.walk(directory, topdown=False):
        for file in files:
            os.remove(os.path.join(root, file))
        for dir in dirs:
            os.rmdir(os.path.join(root, dir))
    os.rmdir(directory)

def custom_warning_handler(message, category, filename, lineno, file=None, line=None):
    if "ffmpeg_reader.py" in filename:
        log_event("FFmpeg Warning", filename, str(message), conn)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Compress files in a directory individually and save metadata to a database.")
    parser.add_argument("--src_directory", help="Source directory to compress files from", 
                        default=os.getenv('SOURCE_DIRECTORY'))
    parser.add_argument("--dest_directory", help="Destination directory to save compressed files",
                        default=os.getenv('DESTINATION_DIRECTORY'))
    parser.add_argument("--db_path", help="Path to the SQLite database file",
                        default=os.getenv('DB_PATH'))
    parser.add_argument("--azure_container", help="Azure Blob Storage container name", 
                        default=os.getenv('AZURE_CONTAINER'))
    parser.add_argument("--azure_connection_string", help="Azure Blob Storage connection string", 
                        default=os.getenv('AZURE_CONNECTION_STRING'))
    
    args = parser.parse_args()
    
    start_time = datetime.now()
    print(f"Process started at: {start_time}")
    
    original_size = get_directory_size(args.src_directory)
    
    # Generate a timestamp
    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
    
    # Open a single database connection
    conn = sqlite3.connect(args.db_path)
    
    # Set custom warning handler
    warnings.showwarning = custom_warning_handler
    
    try:
        # Compress files and save metadata to the database
        compress_files_and_save_to_db(args.src_directory, args.dest_directory, conn, timestamp)
        
        compressed_size = get_directory_size(args.dest_directory)
        
        end_time = datetime.now()
        print(f"Process ended at: {end_time}")
        print(f"Total duration: {end_time - start_time}")
        print(f"Total original size: {format_size(original_size)}")
        print(f"Total compressed size: {format_size(compressed_size)}")
        
        print(f"Directory tree saved to database: {args.db_path}")
        
        # Upload to Azure Blob Storage if specified
        if args.azure_container and args.azure_connection_string:
            upload_to_azure(args.azure_container, args.azure_connection_string, args.dest_directory, timestamp, conn)
            print(f"Files uploaded to Azure Blob Storage container: {args.azure_container}")
            
            # Remove the created destination files and directory
            remove_directory(args.dest_directory)
            print(f"Removed destination directory: {args.dest_directory}")
    finally:
        conn.close()
        warnings.showwarning = warnings._showwarnmsg_impl  # Reset warning handler to default